Cpu hotplug

bryteise opened the pull request:
Resolves https://github.com/cloud-hypervisor/libvirt/issues/3

bryteise committed:
ch: Add helper function for PUT requests

bryteise committed:
ch: Add monitor portion of resize cpu.

Use the newly introduced virCHMonitorPut and add logic to build the
json for Cloud-Hypervisor.

bryteise committed:
ch: Switch ch_monitor to use glib memory management

bryteise committed:
ch: Add domain validation function for vcpu info

Once hotplug support is available we need to be able to validate the
information read from the system matches our expectations. In this
case, we check that thread ids for cpus exist if and only if a cpu is
online.

bryteise committed:
ch: Add CPU hotplug support code

Add functions for managing CPU hotplug. This is intended to be used by
the driver to fully enable CPU hotplug support.

bryteise committed:
ch: Enable CPU hotplug support

Enable the use setvcpus for the cloud-hypervisor domain. Specifically
supporting 'virsh setvcpus --live'.

bryteise commented:
@devimc If you want to look it over

vineethrp commented on src/ch/ch_monitor.h:
```diff
@@ -30,6 +30,7 @@
 #define URL_VMM_SHUTDOWN "vmm.shutdown"
 #define URL_VMM_PING "vmm.ping"
 #define URL_VM_CREATE "vm.create"
+#define URL_VM_RESIZE "vm.resize"
```
Minor nit: This might be better suited in the next commit

bryteise replied:
Ah indeed, fixing.

vineethrp commented on src/ch/ch_monitor.c:
```diff
@@ -1530,6 +1559,17 @@ virCHMonitorPingVMM(virCHMonitorPtr mon)
+                      unsigned int nvcpus)
+{
+    g_autofree char *payload = NULL;
+    if (virCHMonitorBuildResizeJson(&nvcpus, &payload))
```
Do we need to pass nvcpus as a pointer? I don't see the value being modified in any of the functions above in the stack.

bryteise replied:
Ah thanks for pointing that out. I was considering something else alongside doing the memory hotplug and the pointer was a leftover from that design. Fixing.

vineethrp commented on src/ch/ch_hotplug.c:
```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    if (curvcpus != nvcpus) {
```
I think we can move this check to the start of the function by modifying `if (nvcpus < 1 || nvcpus > maxvcpus) { error }`. The current implementation seems to be okay to hot-unplug all vcpus. I feel that we should not allow unplugging all vcpus. This check will also take care of that.

bryteise replied:
The qemu version of this is a bit different but I did try to keep the same overall structure hence the logic around unplugging all cpus not handled here. This check's avoids adding extra inconsistency in case we had a system where `vcpu->online` is true for more vcpus than what `virDomainDefGetVcpus` gives us (the final total of cpus to be onlined at the end of the operation matches the expected value).

Maybe all we should care about is that the min and max boundaries are handled though which I'm open to so I'll update the code for just the simple check against `1` and `maxvcpus`. Cloud-Hypervisor doesn't really handle individual vcpu offlining. Letting it run and see if the vcpu structure updates after it runs corrects the previously mentioned potential inconsistencies seems like a more proactive approach anyway.

So talked myself into fixing =).

vineethrp commented on src/ch/ch_hotplug.c:
```diff
@@ -0,0 +1,194 @@
+        vcpuinfo->online = true;
+    }
+
+    if (virCHMonitorRefreshThreadInfo(mon) < 0)
```
I feel that we need not have this logic here. When there is a change in the vcpu threads, we might get a monitor event and the monitor thread might take care of it. Could you please give it a  try?

vineethrp commented on src/ch/ch_hotplug.c:
```diff
@@ -0,0 +1,194 @@
+        vcpuinfo->online = true;
+    }
+
+    if (virCHMonitorRefreshThreadInfo(mon) < 0)
+        goto error;
+
+    if (virCHProcessSetupVcpus(vm) < 0)
+        goto error;
+
+    if (virCHDomainValidateVcpuInfo(vm) < 0)
+        goto error;
```
I feel that we need not have this logic here. When there is a change in the vcpu threads, we might get a monitor event and the monitor thread might take care of it. Could you please give it a try? If it doesn't work, it should be a bug and needs to be fixed I think. Also, I think We can move `virCHDomainValidateVcpuInfo` check to `virCHProcessSetupThreads` or `virCHProcessSetupVcpus`

bryteise replied:
I'm sorry I am not quite sure what the change you are looking for here is. The first part of your comment I took to removing the `virCHDomainValidateVcpuInfo` function entirely initially. But then reading your comment about moving the `virCHDomainValidateVcpuInfo` function I think maybe you mean removing the `virCHMonitorRefreshThreadInfo` and `virCHProcessSetupVcpus` calls. 

I'm hesitant remove those two calls because if they fail, the failure would be caught at the point of hotplug execution here. Otherwise the failure would be caught sometime later in a more asynchronous check that could be harder to associate with the hotplug api call.

Thoughts?

vineethrp replied:
Sorry for not being clear. I was mentioning that `virCHMonitorRefreshThreadInfo` and `virCHProcessSetupVcpus` are being done in `virCHProcessSetupThreads` whenever we get a lifecycle event about vm or vcpu. So it makes sense to let the monitor thread take care of this update and to avoid duplicating it. You have a valid point that doing it synchronously will help to catch Hotplug errors. But there is always a chance for a race here: If we call `virCHMonitorRefreshThreadInfo` before the threads are populated in procfs, then we get a false alarm here. We would not be having a race in the monitor event handler.

I am fine with duplicating the logic here. Maybe you can add `virCHDomainValidateVcpuInfo` to the `virCHProcessSetupThreads` as well so that we validate the vcpu thread information every time there is a change.

bryteise replied:
Thinking a little more about this, the `virCHDomainHotplugSetVcpus` sets the expected vcpus online (which it knows via `nvcpus`). Then `virCHDomainValidateVcpuInfo` is called to ensure the tids that were read from proc are attached to those newly marked online cpus (and only the marked online cpus). If I add `virCHDomainValidateVcpuInfo` to `virCHProcessSetupThreads`, it would use the last set cpu online values in the `vm->def` which as long as it was from the corresponding hotplug would be fine but isn't that racy too?

It seems we could potentially always be subject to some levels of race condition around when proc updates even if we only read after cloud-hypervisor sends us a message to check.

In the hopes of avoiding stacking potential races, I'd not duplicate the calls and just either move the update and validation to the `virCHProcessSetupThreads` call or keep it as is. Should the `virCHProcessSetupThreads` also be making a `virDomainObjSave` call (and maybe remove the call from `virCHDomainHotplugSetVcpus`)?

vineethrp replied:
I agree about not duplicating the code. And it makes sense to call `virDomainObjSave` in `virCHProcessSetupThreads`. So, I feel we have two options:
1. Remove lines 120 to 130, add `virDomainObjSave` to `virCHProcessSetupThreads` and call `virCHProcessSetupThreads` here in this `virCHDomainHotplugSetVcpus`
2. Remove lines 120 to 130, add `virDomainObjSave` to `virCHProcessSetupThreads` and rely on the event monitor thread to update the vcpu pid information.

Either of the above  are fine with me.

bryteise replied:
Went with option 2, though as noted in my last commit there is an issue with detecting removing vcpus.

vineethrp commented on src/ch/ch_hotplug.c:
```diff
@@ -0,0 +1,194 @@
+ * Returns the true if the nvcpus can be set, false if not.
+ */
+static bool
+virCHDomainSelectHotplugVcpuEntities(virDomainDefPtr def,
```
nit: I think this function validates if the nvcpus is with in allowable ranges. Maybe `virCHDomainValidateHotplugVcpuEntities` be a better name?

bryteise replied:
Yea, I think I was trying to hard to keep the qemu naming convention for similar functions. Fixing.

bryteise committed:
ch: Handle resize events

When resize events are found in the monitor, try to update the thread
id mapping (if a change is detected the domain.xml will also be
updated).

On vm resizes removing vcpus, the event will be triggered but because
cloud-hypervisor does not kill the associated thread right away, the
thread of the removed vcpu is going to still be detected and so
validating the event was processed successfully is skipped for now.

vineethrp commented on src/ch/ch_hotplug.c:
```diff
@@ -0,0 +1,140 @@
+{
+    int ret = -1;
+
+    if (def && nvcpus > virDomainDefGetVcpusMax(def)) {
```
minor: We do this check in `virCHDomainValidateHotplugVcpuEntities`. Probably refactor this error message to `virCHDomainValidateHotplugVcpuEntities` so that we can avoid this check here?

bryteise replied:
Ah good catch, fixing.

vineethrp reviewed (approved):
Nice work! One minor comment, but other than that - Looks to good to me!
